kjdoc/tour.md

```diff
@@ -538,13 +538,440 @@ This section describes KJ APIs that control process execution and low-level inte
+    return boopEvery5SecondsLoop(timer);
+  });
+}
+```
+
+Another possible fix would be to make sure the recursive continuation and then error handler are passed to the same `.then()` invocation:
+
+```c++
+kj::Promise<void> boopEvery5Seconds(kj::Timer& timer) {
+  // No more memory leaks, but hard to reason about.
+  return timer.afterDelay(5 * kj::SECONDS).then([&timer]() {
+    boop();
+  }).then([&timer]() {
+    // Loop by recursing.
+    return boopEvery5Seconds(timer);
+  }, [](kj::Exception&& exception) {
+    // Oh no, an error! Log it and end the loop.
+    KJ_LOG(ERROR, exception);
+    kj::throwFatalException(kj::mv(exception));
+  });
+}
+```
+
+Notice that in this second case, the error handler is scoped so that it does _not_ catch exceptions thrown by the recursive call; it only catches exceptions from `boop()`. This solves the problem, but it's a bit trickier to understand and to ensure that exceptions can't accidentally slip past the error handler.
+
+### Forking and splitting promises
+
+As mentioned above, `.then()` and similar functions consume the promise on which they are called, so can only be called once. But, what if you want to start multiple tasks using the result of a promise? You could solve this in a convoluted way using adapted promises, but KJ has a built-in solution: `.fork()`
+
+```c++
+kj::Promise<int> promise = ...;
+kj::ForkedPromise<int> forked = promise.fork();
+kj::Promise<int> branch1 = promise.addBranch();
+kj::Promise<int> branch2 = promise.addBranch();
+kj::Promise<int> branch3 = promise.addBranch();
+```
+
+A forked promise can have any number of "branches" which represent different consumers waiting for the same result.
+
+Forked promises use reference counting. The `ForkedPromise` itself, and each branch created from it, each represents a reference to the original promise. The original promise will only be canceled if all branches are canceled and the `ForkedPromise` itself is destroyed.
+
+Forked promises require that the result type has a copy constructor, so that it can be copied to each branch. (Regular promises only require the result type to be movable, not copyable.) Or, alternatively, if the result type is `kj::Own<T>` -- which is never copyable -- then `T` must have a method `kj::Own<T> T::addRef()`; this method will be invoked to create each branch. Typically, `addRef()` would be implemented using reference counting.
+
+Sometimes, the copyable requirement of `.fork()` can be burdensome and unnecessary. If the result type has multiple components, and each branch really only needs one of the components, then being able to copy (or refcount) is unnecessary. In these cases, you can use `.split()` instead. `.split()` converts a promise for a `kj::Tuple` into a `kj::Tuple` of promises. That is:
+
+```c++
+kj::Promise<kj::Tuple<kj::Own<Foo>, kj::String>> promise = ...;
+kj::Tuple<kj::Promise<kj::Own<Foo>>, kj::Promise<kj::String>> promises = promise.split();
+```
+
+### Threads
```

Comment:
I don't think this doc needs to go into any detail on it since the header is pretty concise, but it seems worth mentioning that `kj::Thread` provides a primitive for creating a new thread.

Reply:
`kj::Thread` is discussed earlier in the doc. It's not really directly related to promises. If I call it out specifically here, then I also have to explain that there's nothing special about `kj::Thread`-created threads and you can use some other threading library if you prefer, but then that all seems too verbose... so I think I just won't name it here.